xdf.cpp

```diff
@@ -176,251 +177,91 @@ int Xdf::load_xdf(std::string filename)
-                            float data;
-                            Xdf::readBin(file, &data);
-                            streams[index].time_series[v].emplace_back(data);
-                        }
+                      Xdf::readBin(file, &ts);
+                      streams[index].time_stamps.emplace_back(ts);
                     }
-                }
-                else if (streams[index].info.channel_format.compare("double64") == 0)
-                {
-                    //if the time series is empty
-                    if (streams[index].time_series.empty())
-                        streams[index].time_series.resize(streams[index].info.channel_count);
-
-                    //for each sample
-                    for (size_t i = 0; i < numSamp; i++)
+                    else
                     {
-                        //read or deduce time stamp
-                        auto tsBytes = readBin<uint8_t>(file);
-
-                        double ts;  //temporary time stamp
-
-                        if (tsBytes == 8)
-                        {
-                            Xdf::readBin(file, &ts);
-                            streams[index].time_stamps.emplace_back(ts);
-                        }
-                        else
-                        {
-                            ts = streams[index].last_timestamp + streams[index].sampling_interval;
-                            streams[index].time_stamps.emplace_back(ts);
-                        }
-
-                        streams[index].last_timestamp = ts;
+                      ts = streams[index].last_timestamp + streams[index].sampling_interval;
+                      streams[index].time_stamps.emplace_back(ts);
+                    }
+                    
+                    streams[index].last_timestamp = ts;
+                    
+                    if(streams[index].info.channel_format.compare("string") == 0) {
+                      
+                      for(int v = 0; v < streams[index].info.channel_count; ++v) {
+                        auto length = Xdf::readLength(file);
+                        char* buffer = new char[length + 1];
+                        file.read(buffer, length);
+                        buffer[length] = '\0';
 
-                        //read the data
-                        for (int v = 0; v < streams[index].info.channel_count; ++v)
-                        {
+                        streams[index].time_series[v].emplace_back(buffer);
+                      }
+                    } else {
+                      //read the data
+                        if(streams[index].info.channel_format.compare("float32") == 0) {
+                          for (int v = 0; v < streams[index].info.channel_count; ++v)
+                          {
+                          float data;
+                          Xdf::readBin(file, &data);
+                          streams[index].time_series[v].emplace_back(data);
+                          }
```

Comment:
Would be better to extract a function to re-use code (inside an anonymous namespace):

```

namespace {

// Reads data of type T up to `channel_count` and appends
// to `stream.time_series`.
templace <typename T>
void append_data(Stream& stream) {
  for (int v = 0; v < stream.info.channel_count; ++v) {
    T data;
    Xdf::readBin(file, &data);
    stream.time_series[v].push_back(std::move(data));
  }
}

} // namespace

Xdf::Xdf ...
```

and use it like

```
append_data<double>(streams[index]);